### PR TITLE
refactor consumers: one process per stream handler, one stream to many handlers

### PR DESCRIPTION
**Pull request by andycmaj**
wanted to get to multiple consumers per topic so we could do things like...
```
BOTANY_ACTIVITY ----> nudger
                |
                ----> metrics processor
```

 i wasn't happy with the one consumer process having to route all incoming topic messages... introduced a single point of failure in a system that is meant to de-couple asynchronous operations. this allows a many:many relationship between topics and consumers.

```bash
Usage: index.ts [options]

Options:
  -d, --debug              output extra debugging
  -t, --topic <topic>      topic to consume
  -h, --handler <handler>  handler module
  -h, --help               output usage information
```

### local

```bash
# now you run this for consumers
yarn dev:consumers --topic raw_activity --handler enrichRawActivity
```

### deployed

* added a new deployment for each handler
```
 D infrastructure/stream-consumer/templates/deployment.yaml
?? infrastructure/stream-consumer/templates/deployment-enrichRawActivity.yaml
?? infrastructure/stream-consumer/templates/deployment-respondToSlack.yaml
```

### CODE REVIEW

**Comment by andycmaj** [comment]
@davemeyer, @krichdev  how do you feel about this direction?

File: consumers/src/handlers/triggerNudges/index.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { TopicMessageHandler } from '..';
+import CloudEvent from 'shared/cloudEvent';
+
+const handler: TopicMessageHandler = async (cloudEvent: CloudEvent) => {
+  console.log('type', cloudEvent.type);
+
+  // @dmeyer: determine and fire nudges here?
```
Comment (andycmaj) [inline consumers/src/handlers/triggerNudges/index.ts]: @davemeyer this is where you'd fill in the blanks..

```
- [ ]  where should nudge decision happen (should this activity trigger a nudge?)?
    - Alternatives
        - **Enrich**: when we enrich the raw event with an Actor, also determine whether a nudge should fire for the Actor for this Activity.
        - **Process**: look at each activity event (actor + activity) and determine: for the **actor**, should this **activity** trigger a nudge?
```

https://www.notion.so/asynchronous/Streams-and-Events-Implementation-5c83c51913c74e73b41f0b88bbd26302#50de7c7655ad47bd831c3da89453a20f

**Comment by davemeyer** [comment]
if i understand it correctly, it makes sense. so the basic idea is that we are using fairly broad all-encompassing topics. just a couple actual distinct values. then there will be multiple consumers listening for events and only acting on the specific sub-types of the generic topics.

**Comment by andycmaj** [comment]
> only acting on the specific sub-types of the generic topics.

i'm not necessarily thinking that different consumers would act on different sub-types, but rather on specific individual EVENTS that are relevant to that handler-type...

example:

raw event: `github:pull_request:opened`
stream: `raw_events`
consumers: `enrichRawActivity` -> produces `code_request_created`

botany activity: `code_request_created`
stream: `botany_events`
consumers: 
  * `triggerNudges` -> look at each event and determine based on actor/event data if a nudge should be triggered, if so, produce to `nudges`
  * `processActivityMetrics`-> look at each event/actor pair and determine if metrics need to be logged.

**Comment by davemeyer** [comment]
gotcha. well let's diagram that out and make sure we're on the same page with how to structure things going forward

**Comment by andycmaj** [comment]
@davemeyer does this make sense?

ScreenCapture at Thu Aug 22 12:45:37 PDT 2019

**Pull request merged**